File: CS/AsyncChildNodesSelector/AsyncChildNodesSelector.cs
using DevExpress.Xpf.Grid;
using System;
using System.Collections;
using System.Threading;
using System.Threading.Tasks;

namespace AsyncChildNodesSelector {
    public class CustomChildrenSelector : IAsyncChildNodesSelector {
        public Task<bool> HasChildNode(object item, CancellationToken token) {
            return Task.Run(async () => {
                await Task.Delay(250);
                return !(item is StageTask);
            });
        }

        public IEnumerable SelectChildren(object item) {
            throw new NotImplementedException();
        }

        public Task<IEnumerable> SelectChildrenAsync(object item, CancellationToken token) {
            return Task.Run(async () => {
                await Task.Delay(1000);
                return SelectChildNodes(item);
            });
        }

        public IEnumerable SelectChildNodes(object item) {
            if (item is ProjectStage)
                return (item as ProjectStage).Tasks;
            else if (item is ProjectObject)
                return (item as ProjectObject).Stages;
            return null;
        }
    }
}

File: VB/AsyncChildNodesSelector/AsyncChildNodesSelector.vb
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Collections
Imports System.Threading
Imports System.Threading.Tasks

Namespace AsyncChildNodesSelector

    Public Class CustomChildrenSelector
        Implements IAsyncChildNodesSelector

        Public Function HasChildNode(ByVal item As Object, ByVal token As CancellationToken) As Task(Of Boolean) Implements IAsyncChildNodesSelector.HasChildNode
            Return Task.Run(Async Function()
                For i As Integer = 0 To 10 - 1
                    token.ThrowIfCancellationRequested()
                    Await Task.Delay(25)
                Next

                Return Not(TypeOf item Is StageTask)
            End Function)
        End Function

        Public Function SelectChildren(ByVal item As Object) As IEnumerable Implements IChildNodesSelector.SelectChildren
            Throw New NotImplementedException()
        End Function

        Public Function SelectChildrenAsync(ByVal item As Object, ByVal token As CancellationToken) As Task(Of IEnumerable) Implements IAsyncChildNodesSelector.SelectChildrenAsync
            Return Task.Run(Async Function()
                For i As Integer = 0 To 10 - 1
                    token.ThrowIfCancellationRequested()
                    Await Task.Delay(100)
                Next

                Return SelectChildNodes(item)
            End Function)
        End Function

        Public Function SelectChildNodes(ByVal item As Object) As IEnumerable
            If TypeOf item Is ProjectStage Then
                Return TryCast(item, ProjectStage).Tasks
            ElseIf TypeOf item Is ProjectObject Then
                Return TryCast(item, ProjectObject).Stages
            End If

            Return Nothing
        End Function
    End Class
End Namespace
